lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -629,159 +629,177 @@ This is problematic for linalg ops for a few reasons:
 h! Although it is a bit like using a hammer to paint, our workaround is to use
 tensor.extract to access the elements of the input tensor inside our linalg
 generic op's payload.
-
-Current TODO's:
-    1. gather most of the boilerplate out of this op and make it into an
-adaptive pooling helper function.
-    2. figure out what to do with the conflicting decompositions in
-DecomposeComplexOps.cpp
-    3. Implement more efficient passes for when the kernel-size, input spatial
-dims, and output spatial dims are constant.
 */
 
 namespace {
-class ConvertAtenAdaptiveAvgPool1dOp
-    : public OpConversionPattern<AtenAdaptiveAvgPool1dOp> {
+
```

Comment:
My understanding of the design is that we have AdaptivePoolingHelper, which is inherited by an AdaptiveMaxPoolingHelper and an AdaptiveAvgPoolingHelper, each of which gets templated to the 1-3d case.

The code was a little hard to read for me and I wonder if there's an easier way to do this.

Reply:
I don't know enough to comment on the tradeoff between:
1. code duplication
2. somebody reading this code having to jump around due to inheritance and templating
but generally in this codebase i see a lot of places where duplication is preferred over inheritance / templating for cases like 1d-3d. E.g. the conv1d-3d ops.

Reply:
> I don't know enough to comment on the tradeoff between:
> 
>     1. code duplication
> 
>     2. somebody reading this code having to jump around due to inheritance and templating
> 
> 
> but generally in this codebase i see a lot of places where duplication is preferred over inheritance / templating for cases like 1d-3d. E.g. the conv1d-3d ops.

That is a good question. Do you think simply adding some more/better commentary in the code would make this more tractable? 

> My understanding of the design is that we have AdaptivePoolingHelper, which is inherited by an AdaptiveMaxPoolingHelper and an AdaptiveAvgPoolingHelper, each of which gets templated to the 1-3d case.
> 
> The code was a little hard to read for me and I wonder if there's an easier way to do this.

Yeah, I found out I left in something which probably makes it a bit more complicated for no reason, so let me make a change and add some better comments. 

Before this, I tried a few different implementations, and felt like this was probably a smidge less readable than some other options, but much easier to adapt in the future if need be.